### PR TITLE
Python: Provide version ranges for each dependency

### LINKED ISSUE #7377: Python: Provide version ranges for each dependency
### Feature Request / Improvement

To follow up on #6620, I am using `pyiceberg` (heavily) for a library that I'm developing. Unfortunately, `pyiceberg` has very restrictive dependencies, and they are in turn determining the bounds for our dependencies (even though I have specified my bounds very loosely). This is problematic, because my users will install my library alongside others in the same environment, so it should be as compatible as possible.

Currently, using a newer version of certain dependencies (e.g. say I want to use `s3fs==2023.4.0`) requires updating the bound in the library and creating a new release. This isn't very sustainable, as `pyiceberg` would need to release any time one of its dependencies adds a new version to keep up with the latest and greatest, nor is it proactive. We can mitigate this by using relatively broad ranges, especially on the upper end.

There's no real risk to doing this, because consumers can always further restrict/pin dependencies (but they can't loosen restrictive bounds by libraries).

There are very valid arguments against capping dependencies (see https://iscinumpy.dev/post/bound-version-constraints/ for a great post on the topic). However, even providing/relaxing the ranges would be a great start!

### Query engine

None

### PR DESCRIPTION
**Pull request by deepyaman**
Closes #7377 

I have gone through each of the dependencies one by one and relaxed them, in most cases testing with a version at the bottom and top of the range to make sure it's not relaxed too far. I have also tried to check library usage (downloads) for older version and security risks, where applicable.

I've added temporary comments marked `# JUST:` inline, with justifications of the new bounds. They should be removed before merging.

This doesn't fully uncap any dependencies (although I would be happy to do that, too; starting off more conservative :)).

Finally, this fixes missing `zstandard` extra; see `ImportError`s like the below, even though the extra isn't defined as far as I can see: https://github.com/apache/iceberg/blob/da37818bc82aee379dbcd913f73d8ee75fccc63c/python/pyiceberg/avro/codecs/zstandard_codec.py#L49

### CODE REVIEW

File: python/pyproject.toml
Comment (Fokko) [inline python/pyproject.toml:47]: Can you make sure that all the bound also have upper bounds?
```suggestion
requests = "^2.20,<3" # JUST: Tests pass with 2.22 (lower versions aren't compatible with some other strictly-pinned dependencies); picked nearby 2.20 as the most recent version without security risks.
```

**Comment by Fokko** [comment]
@deepyaman This makes sense to me. I have one comment on adding upper bounds so we don't upgrade to the next major release without knowing it.

File: python/pyproject.toml
Comment (deepyaman) [reply python/pyproject.toml:47]: They already do! Caret requirements work like this by default: https://python-poetry.org/docs/dependency-specification/#caret-requirements

**Comment by corleyma** [comment]
Setting upper bounds has a lot of problems for library code that is intended to be used by a lot of third parties.  See also[ this blog](https://iscinumpy.dev/post/bound-version-constraints/) for a detailed discussion of the problems this creates.  Tl;dr for some of the big points:

It would be better to leave upper versions unbound, and to pin around broken versions if/when they are discovered.  
 - Your library users can always manually drop down to a lower version of a dependency if they encounter an issue
 - However, if you pin upper bound, users have no way of adopting a newer version of the dependency even if there is no actual incompatibility--they have to wait for you to release a new version without the pin.
 - Since many of your dependencies are core members of the python data ecosystem, and given how long PRs tend to sit in this project currently, it's almost certain that you will be holding folks back from adopting the latest and greatest when e.g. pyarrow drops a new major version (happens every few months).
 
 There are many other important problems outlined in the blog post, but these above should be enough to argue convincingly for setting minimum versions only, and updating your pins when genuine incompatibility is discovered.

**Comment by deepyaman** [comment]
> Setting upper bounds has a lot of problems for library code that is intended to be used by a lot of third parties. See also[ this blog](https://iscinumpy.dev/post/bound-version-constraints/) for a detailed discussion of the problems this creates. Tl;dr for some of the big points:
> 
> It would be better to leave upper versions unbound by default, and to pin around broken versions if/when they are discovered.
> 
> * Your library users can always manually drop down to a lower version of a dependency if they encounter an issue
> * However, if you pin upper bound, users have no way of adopting a newer version of the dependency even if there is no actual incompatibility--they have to wait for you to release a new version without the pin.
> * Since many of your dependencies are core members of the python data ecosystem, and given how long PRs tend to sit in this project currently, it's almost certain that you will be holding folks back from adopting the latest and greatest when e.g. pyarrow drops a new major version (happens every few months).
> 
> There are many other important problems outlined in the blog post, but these above should be enough to argue convincingly for setting minimum versions only, and updating your pins when genuine incompatibility is discovered.

@corleyma I largely (completely?) agree, and even referenced the same article in the issue I created for this. 😄 

> There are very valid arguments against capping dependencies (see https://iscinumpy.dev/post/bound-version-constraints/ for a great post on the topic). However, even providing/relaxing the ranges would be a great start!

But, as I also mention, it's a huge win to even relax them this significantly in my books. I figured a PR to relax upper bounds would be less contentious to start with. :)

**Comment by deepyaman** [comment]
> > @deepyaman This makes sense to me. I have one comment on adding upper bounds so we don't upgrade to the next major release without knowing it.
> 
> Thanks again for the review! If this looks good, I will go ahead and remove the temporary JUST comments later today.

@Fokko Cleaned up and should be good to go! 🤞

File: python/pyproject.toml
Comment (Fokko) [reply python/pyproject.toml]: Thanks, I learned something today. Some context: I was suggesting this because I like Dependabot to bump the versions: https://github.com/apache/iceberg/pull/7309/files#diff-1a45bd938160f7761769e7b6fbc819d1cc7f84ce37f1a6a23ef83cc4e214abc8L56 This way we can dependabot create PRs every week, and see if we're still compatible with the latest version.

Comment (deepyaman) [reply python/pyproject.toml]: I'm happy to change the syntax if you think it will help with Dependabot, but I expect (with the loosened dependencies) that you'll update the upper bound much less frequently.

Comment (Fokko) [reply python/pyproject.toml]: @deepyaman That would be great! Having fewer notifications is good, as long as we're being notified of the major changes, so we can check if there are any API changes that break compatibility.

File: python/pyproject.toml
Comment (Fokko) [inline python/pyproject.toml:62]: I think this should be in line with fsspec:
```suggestion
s3fs = { version = ">=2021.09.0,<2024.1.0", optional = true } # Upper bound set arbitrarily, to be reassessed in early 2024.
```

File: python/pyproject.toml
Comment (Fokko) [inline python/pyproject.toml:63]: I think this should be in line with fsspec:
```suggestion
adlfs = { version = ">=2021.09.0,<2024.1.0", optional = true } # Upper bound set arbitrarily, to be reassessed in early 2024.
```

File: python/pyproject.toml
Comment (Fokko) [inline python/pyproject.toml:93]: Nice! Great catch!

File: python/pyproject.toml
Comment (Fokko) [inline python/pyproject.toml:61]: FYI. This one is hard pinned by `aiobotocore`: https://github.com/aio-libs/aiobotocore/blob/master/setup.py#L10

**Review by Fokko — COMMENTED** [review]
> It would be better to leave upper versions unbound by default, and to pin around broken versions if/when they are discovered.

My question would be. How would you discover this? I like to leverage Dependabot for this.

> Your library users can always manually drop down to a lower version of a dependency if they encounter an issue

Of course, but it is not a nice experience and maybe not trivial for everyone. It would be nice to have some kind of reproducibility. It can be that the CI starts failing because some major version has been released. In the Python community, we have seen this many times. I agree that we don't should lock on a minor version, but keeping it completely open feels to me like asking for problems.

I would suggest pinning it on major versions and losening up the minor ones as already done by @deepyaman

File: python/pyproject.toml
Comment (deepyaman) [reply python/pyproject.toml]: Done!

File: python/pyproject.toml
Comment (deepyaman) [reply python/pyproject.toml]: Sounds good; updated! Maybe it passes the unit tests but doesn't actually work, so I've bumped it as you've recommended.

File: python/pyproject.toml
Comment (deepyaman) [reply python/pyproject.toml:62]: I previously added a comment as part of the justifications:

> `# TODO: Remove? Definitely doesn't seem to hold true to `adlfs`, and `s3fs` enforces it itself.`

To expand upon that, if you install `s3fs`, it will automatically bound the version of `fsspec` (in newer versions). On the other hand, `adlfs` does not release in lockstep with `fsspec`; e.g. the latest version there is `2023.1.0`, and its `fsspec` requirements is just `fsspec>=2021.10.1`.

Comment (Fokko) [reply python/pyproject.toml:62]: I think removing would be better then.

File: python/pyproject.toml
Comment (Fokko) [inline python/pyproject.toml:50]: Should we clean up these justifications as well?

**Review by Fokko — APPROVED** [review]
LGTM, Two small nits left in the comments.

File: python/pyproject.toml
Comment (deepyaman) [reply python/pyproject.toml:62]: Oh, sorry, rereading my comment I see I wasn't clear. The "Remove?" was regarding the comment saying the versions should be in lockstep; all of the dependencies still need to be kept (`fsspec` as a required dependency for the library with appropriate bounds, and the two extras for different filesystems).

File: python/pyproject.toml
Comment (deepyaman) [reply python/pyproject.toml:50]: I left this one as a reasonable answer as to why the 5.4.0 lower bound shouldn't be reduced (should somebody want to reevaluate the bounds later), but I'm happy to remove this comment if preferred. I cleaned up the more nebulous justifications. Your call on this.

**Comment by Fokko** [comment]
I think this is a great first step! Thanks for working on this @deepyaman and @corleyma for the insights

**Pull request merged**
